common/flatpak-dir.c

```diff
@@ -4748,13 +4748,97 @@ export_desktop_file (const char   *app,
   return ret;
 }
 
+static GStrv
+enumerate_owned_dbus_names (GKeyFile  *metakey,
+                            GError   **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) owned_names = g_ptr_array_new_with_free_func (g_free);
+  g_auto(GStrv) keys = g_key_file_get_keys (metakey,
+                                            FLATPAK_METADATA_GROUP_SESSION_BUS_POLICY,
+                                            NULL,
+                                            &local_error);
+  GStrv iter = keys;
+
+  if (keys == NULL)
+    {
+      /* No group, return an empty strv */
+      if (g_error_matches (local_error,
+                           G_KEY_FILE_ERROR,
+                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND))
+        {
+          g_ptr_array_add (owned_names, NULL);
+          return (GStrv) g_ptr_array_free (owned_names, FALSE);
+        }
+
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+
+  for (; *iter != NULL; ++iter)
+    {
+      g_autofree char *value = g_key_file_get_string (metakey,
+                                                      FLATPAK_METADATA_GROUP_SESSION_BUS_POLICY,
+                                                      *iter,
+                                                      error);
+
+      if (value == NULL)
+        return NULL;
+
+      if (g_strcmp0 (value, "own") == 0)
+        g_ptr_array_add (owned_names, g_strdup (*iter));
+    }
+
+  g_ptr_array_add (owned_names, NULL);
+  return (GStrv) g_ptr_array_free (g_steal_pointer (&owned_names), FALSE);
+}
+
+typedef gboolean (*MultiplePrefixesComparisonFunc) (const char         *path,
+                                                    const char * const *prefixes);
+
+static GStrv
+get_permissible_prefixes (GKeyFile                        *metakey,
+                          const char                      *source_path,
+                          const char                      *app_id,
+                          MultiplePrefixesComparisonFunc  *out_match_prefixes_func,
+                          GError                         **error)
+{
+  g_autoptr(GPtrArray) prefixes = NULL;
+
+  g_return_val_if_fail (out_match_prefixes_func != NULL, FALSE);
+
+  prefixes = g_ptr_array_new_with_free_func (g_free);
+
+  g_ptr_array_add (prefixes, g_strdup (app_id));
+
+  if (g_str_has_suffix (source_path, "share/dbus-1/services"))
```

Comment:
suffix? Shouldn't this be prefix? Also maybe flatpak_has_path_prefix() so its not confused by share/dbus-1/serviceswithsomeotherthing

Reply:
I think this does need to be `g_str_has_suffix`. `source_path` here is the entire path down the traversal tree, we want to know if it ends with share/dbus-1/services so that we apply a different logic in collecting exported files from the directory.

Reply:
Still, it is easy to confuse this by having "share/dbus-1/services" at the end of some path that is not actually *the* /share/dbus-1/services, like some/other/dir/share/dbus-1/services.

Reply:
I see, that would be a suboptimal outcome indeed.

I'll see if there's a way to compute this properly - I think `source_path` is the full path to the source file so just checking on a prefix probably wouldn't work.